### PR TITLE
Implement BlockedBloomFilter

### PR DESCRIPTION
**Pull request by k-jingyang**
Solves #78

### CODE REVIEW

**Comment by k-jingyang** [comment]
Sharing the benchmark on my PC

```
$ cargo bench -- "bloom filter"

     Running benches/bloom.rs (target/release/deps/bloom-2e83c48a73000131)
bloom filter add key    time:   [411.51 ns 412.41 ns 413.53 ns]
                        change: [-6.7615% -5.9719% -5.1901%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  5 (5.00%) high mild
  9 (9.00%) high severe

bloom filter contains key, true positive (1%)
                        time:   [41.095 ns 41.566 ns 42.132 ns]
                        change: [-15.853% -12.870% -9.9857%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  1 (1.00%) high mild
  13 (13.00%) high severe

bloom filter contains key, true positive (0.1%)
                        time:   [48.389 ns 49.241 ns 50.302 ns]
                        change: [-15.548% -11.948% -7.9643%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe

bloom filter contains key, true positive (0.01%)
                        time:   [56.502 ns 57.368 ns 58.473 ns]
                        change: [-17.297% -13.599% -9.7068%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  3 (3.00%) high mild
  11 (11.00%) high severe

bloom filter contains key, true positive (0.0009999999%)
                        time:   [62.768 ns 63.557 ns 64.553 ns]
                        change: [-17.250% -14.014% -10.462%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe

bloom filter add key - blocked bloom filter
                        time:   [408.43 ns 411.52 ns 415.83 ns]
Found 16 outliers among 100 measurements (16.00%)
  5 (5.00%) high mild
  11 (11.00%) high severe

bloom filter contains key, true positive (1%) - blocked bloom filter
                        time:   [35.358 ns 35.803 ns 36.400 ns]
                        change: [-19.207% -15.541% -11.769%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 18 outliers among 100 measurements (18.00%)
  7 (7.00%) high mild
  11 (11.00%) high severe

bloom filter contains key, true positive (0.1%) - blocked bloom filter
                        time:   [43.504 ns 43.904 ns 44.432 ns]
                        change: [-16.757% -13.042% -8.3987%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  2 (2.00%) high mild
  10 (10.00%) high severe

bloom filter contains key, true positive (0.01%) - blocked bloom filter
                        time:   [47.467 ns 48.095 ns 48.989 ns]
                        change: [-16.809% -13.030% -9.5106%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 18 outliers among 100 measurements (18.00%)
  4 (4.00%) high mild
  14 (14.00%) high severe

bloom filter contains key, true positive (0.0009999999%) - blocked bloom filter
                        time:   [50.597 ns 50.977 ns 51.449 ns]
                        change: [-17.290% -13.593% -9.2939%] (p = 0.00 < 0.05)
                        Performance has improved.
```

**Comment by marvin-j97** [comment]
Would you mind basing this on the `3.0.0` branch? I don't plan on adding blocked bloom filters in V2 anyway, and I already refactored the module tree to allow multiple filter types (`src/segment/filter`) in there.

**Comment by k-jingyang** [comment]
> Would you mind basing this on the `3.0.0` branch? I don't plan on adding blocked bloom filters in V2 anyway, and I already refactored the module tree to allow multiple filter types (`src/segment/filter`) in there.

Sure, will do. Thanks!

**Comment by k-jingyang** [comment]
After re-basing off the 3.0.0 branch, here's the bechmarks. Noted that the benchmarking parameters changed compared to my above benchmark, hence the difference in the standard bloom filter benchmarks.

```
$ cargo bench -- ".+ bloom filter"

     Running benches/bloom.rs (target/release/deps/bloom-9b9f9b6ef85c0cd2)
standard bloom filter add key
                        time:   [533.32 ns 535.43 ns 538.80 ns]
                        change: [-5.1674% -3.9522% -2.7998%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 17 outliers among 100 measurements (17.00%)
  3 (3.00%) high mild
  14 (14.00%) high severe

standard bloom filter contains key, true positive (1%)
                        time:   [145.00 ns 145.68 ns 146.52 ns]
                        change: [-2.8564% -0.8346% +1.2524%] (p = 0.43 > 0.05)
                        No change in performance detected.
Found 17 outliers among 100 measurements (17.00%)
  2 (2.00%) high mild
  15 (15.00%) high severe

standard bloom filter contains key, true positive (0.1%)
                        time:   [187.06 ns 192.66 ns 199.79 ns]
                        change: [-5.0150% -3.0765% -0.9774%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe

standard bloom filter contains key, true positive (0.01%)
                        time:   [249.35 ns 250.61 ns 252.20 ns]
                        change: [-2.2533% -1.1636% -0.0419%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 16 outliers among 100 measurements (16.00%)
  2 (2.00%) high mild
  14 (14.00%) high severe

standard bloom filter contains key, true positive (0.0009999999%)
                        time:   [275.90 ns 276.61 ns 277.49 ns]
                        change: [-8.7734% -5.6394% -2.8336%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  3 (3.00%) high mild
  11 (11.00%) high severe

blocked bloom filter add key
                        time:   [498.51 ns 502.83 ns 508.08 ns]
                        change: [-1.3098% -0.2746% +0.7744%] (p = 0.62 > 0.05)
                        No change in performance detected.
Found 16 outliers among 100 measurements (16.00%)
  3 (3.00%) high mild
  13 (13.00%) high severe

blocked bloom filter contains key, true positive (1%)
                        time:   [77.661 ns 78.615 ns 79.697 ns]
                        change: [-6.7851% -3.2400% +0.3533%] (p = 0.08 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

blocked bloom filter contains key, true positive (0.1%)
                        time:   [101.62 ns 102.36 ns 103.20 ns]
                        change: [-6.3367% -3.3385% -0.3909%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 14 outliers among 100 measurements (14.00%)
  6 (6.00%) high mild
  8 (8.00%) high severe

blocked bloom filter contains key, true positive (0.01%)
                        time:   [120.55 ns 124.92 ns 130.13 ns]
                        change: [+2.7154% +6.9422% +11.441%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe

blocked bloom filter contains key, true positive (0.0009999999%)
                        time:   [125.38 ns 130.16 ns 135.27 ns]
                        change: [-0.6506% +2.7169% +5.9467%] (p = 0.10 > 0.05)
                        No change in performance detected.
```

**Comment by k-jingyang** [comment]
The next steps in the PR will be:
1. An AMQFilter trait for use in the segment `src/segment/inner.rs`
2. A generic builder/factory to decode the appropriate filter based on Reader
  - I'm assuming that each filter type can encode different variables, thus decode implementation will have to be provided by each filter type

Please correct me if I'm wrong, or if you have other ideas.

**Comment by marvin-j97** [comment]
Sounds good to me

One more thing is comparing the FPR of the bloom filters. Blocked _should_ have slightly higher FPR.

File: src/segment/filter/mod.rs [hunk 1]
```diff
@@ -38,3 +45,56 @@ impl BloomConstructionPolicy {
...
+    fn bytes(&self) -> &[u8];
+    fn len(&self) -> usize;
+    fn contains(&self, item: &[u8]) -> bool;
+    fn contains_hash(&self, hash: (u64, u64)) -> bool;
+}
+
+pub struct AMQFilterBuilder {}
+
+impl AMQFilterBuilder {
+    pub fn decode_from<R: Read>(reader: &mut R) -> Result<Box<dyn AMQFilter + Sync>, DecodeError> {
```
Comment (k-jingyang) [inline src/segment/filter/mod.rs]: I couldn't use impl Decode, because we're not returning a AMQFilterBuilder here.

Otherwise, the method signature is similar

File: src/segment/filter/standard_bloom/mod.rs [hunk 2]
```diff
@@ -50,20 +93,10 @@ impl Encode for StandardBloomFilter {
...
-        }
-
-        // NOTE: Filter type (unused)
-        let filter_type = reader.read_u8()?;
-        assert_eq!(0, filter_type, "Invalid filter type");
-
+#[allow(clippy::len_without_is_empty)]
+impl StandardBloomFilter {
+    // To be used by AMQFilter after magic bytes and filter type have been read and parsed
+    pub(super) fn decode_from<R: Read>(reader: &mut R) -> Result<Self, DecodeError> {
```
Comment (k-jingyang) [inline src/segment/filter/standard_bloom/mod.rs]: method signature is similar to decode_from in Decode. Changed to it (super) visibility because it's only to be used by AMQFilterBuilder

Comment (k-jingyang) [reply src/segment/filter/standard_bloom/mod.rs]: Although, it feels slightly weird to reuse DecodeError here. Not sure if we should use another type of error instead

File: src/segment/filter/mod.rs [hunk 3]
```diff
@@ -38,3 +45,56 @@ impl BloomConstructionPolicy {
...
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Self::StandardBloom),
+            1 => Ok(Self::BlockedBloom),
+            _ => Err(()),
+        }
+    }
+}
+
+pub trait AMQFilter: Sync + Send {
```
Comment (k-jingyang) [inline src/segment/filter/mod.rs]: Adding Sync + Send here, otherwise we would get errors from `blob_drop_after_flush`

File: src/segment/filter/mod.rs [hunk 4]
```diff
@@ -38,3 +45,56 @@ impl BloomConstructionPolicy {
...
+        match filter_type {
+            FilterType::StandardBloom => StandardBloomFilter::decode_from(reader)
+                .map(Self::wrap_filter)
+                .map_err(|e| DecodeError::from(e)),
+            // TODO: Implement
+            FilterType::BlockedBloom => Err(DecodeError::InvalidHeader("BlockedBloom")),
+        }
+    }
+
+    fn wrap_filter<T: 'static + AMQFilter + Sync>(filter: T) -> Box<dyn AMQFilter + Sync> {
```
Comment (k-jingyang) [inline src/segment/filter/mod.rs]: Tbh, not super clear if this is the right way to do this. Let me know if there's a better way.

File: src/segment/filter/blocked_bloom/mod.rs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    fn encode_into<W: Write>(&self, writer: &mut W) -> Result<(), EncodeError> {
+        // Write header
+        writer.write_all(&MAGIC_BYTES)?;
+
+        writer.write_u8(super::FilterType::BlockedBloom as u8)?;
+
+        // NOTE: Hash type (unused)
+        writer.write_u8(0)?;
+
+        writer.write_u64::<LittleEndian>(self.num_blocks as u64)?;
```
Comment (k-jingyang) [inline src/segment/filter/blocked_bloom/mod.rs:84]: there's an underlying assumption here that each cache line block will always be 64 bytes

File: src/segment/filter/mod.rs (hunk 4 above)
Comment (marvin-j97) [reply src/segment/filter/mod.rs]: Instead of a Boxed dyn object, enum dispatch could be used here.

E.g.

```rs
#[enum_dispatch(AMQFilter)]
enum AnyFilter {
  Standard(StandardBloom),
  Blocked(BlockedBloom),
}
```

File: src/segment/filter/blocked_bloom/mod.rs (hunk 5 above)
Comment (marvin-j97) [reply src/segment/filter/blocked_bloom/mod.rs:84]: Nothing that can be done about that I think.

Most systems use 512 bits; Apple Silicon seems to use 1024 bits, meaning 2 blocks would share a cache line.

File: src/segment/filter/blocked_bloom/mod.rs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    /// Returns `true` if the hash may be contained.
+    ///
+    /// Will never have a false negative.
+    #[must_use]
+    fn contains_hash(&self, (mut h1, mut h2): CompositeHash) -> bool {
+        let block_idx = h1 % (self.num_blocks as u64);
+
+        for i in 1..(self.k as u64) {
+            h1 = h1.wrapping_add(h2);
+            h2 = h2.wrapping_add(i);
```
Comment (marvin-j97) [inline src/segment/filter/blocked_bloom/mod.rs]: This should probably be `wrapping_mul`, I already changed it in the standard filter; not sure how much of a difference it makes, but I think the enhanced double hashing formula is `h(1) + i * h2 + f(i)`

Same in Builder

File: src/segment/filter/blocked_bloom/builder.rs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        ((m / 8.0).ceil() * 8.0) as usize
+    }
+
+    /// Adds the key to the filter.
+    pub fn set_with_hash(&mut self, (mut h1, mut h2): CompositeHash) {
+        let block_idx = h1 % (self.num_blocks as u64);
+
+        for i in 1..(self.k as u64) {
+            h1 = h1.wrapping_add(h2);
+            h2 = h2.wrapping_add(i);
```
Comment (marvin-j97) [inline src/segment/filter/blocked_bloom/builder.rs]: These lines should probably be moved to the end of the loop iteration.

Same for filter (reader)

File: src/segment/filter/blocked_bloom/builder.rs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+        // NOTE: Some sensible minimum
+        let fpr = fpr.max(0.000_001);
+
+        // TODO: m and k is still calculated by traditional standard bloom filter formula
+        let m = Self::calculate_m(n, fpr);
+        let bpk = m / n;
+        let k = (((bpk as f32) * LN_2) as usize).max(1);
+
+        let num_blocks = m.div_ceil(CACHE_LINE_BYTES);
```
Comment (marvin-j97) [inline src/segment/filter/blocked_bloom/builder.rs]: Note that `m` is in *bits*, but `CACHE_LINE_BYTES` is in *bytes*, so currently there are 8x more blocks allocated than necessary.

File: src/segment/filter/blocked_bloom/builder.rs [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+        assert!(bpk > 0);
+        assert!(n > 0);
+
+        let bpk = bpk as usize;
+
+        let m = n * bpk;
+        let k = (((bpk as f32) * LN_2) as usize).max(1);
+
+        let num_blocks = m.div_ceil(CACHE_LINE_BYTES);
```
Comment (marvin-j97) [inline src/segment/filter/blocked_bloom/builder.rs]: Note that `m` is in *bits*, but `CACHE_LINE_BYTES` is in *bytes*, so currently there are 8x more blocks allocated than necessary.

**Review by marvin-j97 — CHANGES_REQUESTED** [review]

File: src/segment/filter/blocked_bloom/builder.rs [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    /// Adds the key to the filter.
+    pub fn set_with_hash(&mut self, (mut h1, mut h2): CompositeHash) {
+        let block_idx = h1 % (self.num_blocks as u64);
+
+        for i in 1..(self.k as u64) {
+            h1 = h1.wrapping_add(h2);
+            h2 = h2.wrapping_mul(i);
+
+            let idx = h1 % (CACHE_LINE_BYTES as u64);
```
Comment (marvin-j97) [inline src/segment/filter/blocked_bloom/builder.rs]: Similarly here; `CACHE_LINE_BYTES` is 64 bytes, we are getting the *bit index* in a block here, so it should be `CACHE_LINE_BYTES * 8`.

File: src/segment/filter/blocked_bloom/builder.rs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+            self.inner
+                .enable_bit(Self::get_bit_idx(block_idx as usize, idx as usize));
+        }
+    }
+
+    pub fn get_bit_idx(block_idx: usize, idx_in_block: usize) -> usize {
+        block_idx * CACHE_LINE_BYTES as usize + idx_in_block
```
Comment (marvin-j97) [inline src/segment/filter/blocked_bloom/builder.rs]: And here the result is a *bit index*, but `block_idx * CACHE_LINE_BYTES` returns the `byte index` of a block, so it should also be `block_idx * CACHE_LINE_BYTES * 8`.

File: src/segment/filter/blocked_bloom/builder.rs (hunk 7 above)
Comment (k-jingyang) [reply src/segment/filter/blocked_bloom/builder.rs]: This was deliberate, to add variance between the choice of block and the first bit set in the block.

An edge case would be if num_of_blocks == cache_line_bytes. This would cause us to always set bit X of block X as the first bit.

File: src/segment/filter/blocked_bloom/builder.rs (hunk 11 above)
Comment (k-jingyang) [reply src/segment/filter/blocked_bloom/builder.rs]: good catch!

File: src/segment/filter/mod.rs (hunk 4 above)
Comment (k-jingyang) [reply src/segment/filter/mod.rs]: Thanks for this. Learnt something new

File: src/segment/filter/mod.rs [hunk 12]
```diff
@@ -38,3 +45,60 @@ impl BloomConstructionPolicy {
...
+}
+
+#[enum_dispatch::enum_dispatch(AMQFilter)]
+#[derive(PartialEq, Debug)]
+pub enum BloomFilter {
+    StandardBloom(StandardBloomFilter),
+    BlockedBloom(BlockedBloomFilter),
+}
+
+pub enum BloomFilterType {
```
Comment (k-jingyang) [inline src/segment/filter/mod.rs:64]: Can't think of a way to not introduce this enum. Could be missing something.

The primary purpose of this is to have an enumeration of filter_type values during decode & encode.

**Comment by k-jingyang** [comment]
Still pending:
1. Measuring FPR for BlockedBloomFilter

File: src/segment/filter/mod.rs (hunk 12 above)
Comment (marvin-j97) [reply src/segment/filter/mod.rs:64]: Yeah we need this

File: src/segment/filter/blocked_bloom/builder.rs (hunk 7 above)
Comment (marvin-j97) [reply src/segment/filter/blocked_bloom/builder.rs]: That is porobably not an issue though because blocks are always logically isolated. It's probably also quite an edge case.

**Comment by k-jingyang** [comment]
I added the FPR test to `tests/bloom_fpr.rs`. May not be required to be in the repo though.

The FPR is much higher than I thought

```
// [Theoretical] FPR: 1.0000%, [Empirical] Standard Bloom FPR: 0.0002, Blocked Bloom FPR: 0.0313%
// [Theoretical] FPR: 0.1000%, [Empirical] Standard Bloom FPR: 0.0000, Blocked Bloom FPR: 0.0303%
// [Theoretical] FPR: 0.0100%, [Empirical] Standard Bloom FPR: 0.0000, Blocked Bloom FPR: 0.0287%
// [Theoretical] FPR: 0.0010%, [Empirical] Standard Bloom FPR: 0.0000, Blocked Bloom FPR: 0.0257%

// n = 5000000, [Empirical] Standard Bloom FPR: 0.0006, Blocked Bloom FPR: 0.0276%
// n = 10000000, [Empirical] Standard Bloom FPR: 0.0000, Blocked Bloom FPR: 0.0108%
// n = 15000000, [Empirical] Standard Bloom FPR: 0.0000, Blocked Bloom FPR: 0.0086%
```

**Comment by marvin-j97** [comment]
Continuing this on the `feat/blocked-bloom` branch
